internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,470 @@
+		template.DNSNames = slices.Clone(idTemplate.DNSNames)
+		template.EmailAddresses = slices.Clone(idTemplate.EmailAddresses)
+	}
+
+	// The identity CA cert is self-signed, which allows a client to verify the
+	// incoming version of this with the pinned version.
+	certDER, err := x509.CreateCertificate(rand.Reader, template, template, signer.Public(), signer)
+	if err != nil {
+		return nil, err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// isCertActive checks of the current time now is within the start and end time
+// validity period as defined by the certificate.
+func isCertActive(cert *x509.Certificate) bool {
+	now := timeNow()
+	// Note that we shorten the NotAfter timestamp by a tlsCertRenewWindow
+	// duration to avoid a TLS handshake race against the expiry time.
+	if now.Before(cert.NotBefore) || now.After(cert.NotAfter.Add(-tlsCertRenewWindow)) {
+		return false
+	}
+	return true
+}
+
+// isCertDerived checks if the public key in the certificate matches
+// the signer public key. This confirms the certificate matches the
+// identity key.
+func isCertDerived(cert *x509.Certificate, signer crypto.Signer) bool {
+	signerPublic := signer.Public().(ed25519.PublicKey)
+	certPublic := cert.PublicKey.(ed25519.PublicKey)
+	return subtle.ConstantTimeCompare(signerPublic, certPublic) == 1
+}
+
+// expectPermission return an error if the specified directory or file
+// path is not matching the supplied permissions.
+func expectPermission(path string, perm fs.FileMode) error {
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	actualPerm := info.Mode().Perm()
+	if actualPerm != perm {
+		return fmt.Errorf("expected permission 0o%o (got 0o%o) for %q", perm, actualPerm, path)
+	}
+	return nil
+}
+
+// pathExists returns true of the path exists, false if it does not. If the
+// operation reports an unrelated error, the error is returned.
+func pathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// defaultCertSubject provides a default Subject Common Name for x509
+// certificates.
+func defaultCertSubject(fingerprint string) pkix.Name {
+	// The common name must not exceed 64 bytes (RFC 5280).
+	var cn strings.Builder
+	// <= 55 bytes for the prefix (program name)
+	for _, r := range cmd.ProgramName {
+		cnLen := len(cn.String())
+		runeLen := len(string(r))
+		if (cnLen + runeLen) > 55 {
+			break
+		}
+		cn.WriteRune(r)
+	}
+	// 1 byte for the separator.
+	cn.WriteString("-")
+	prefixLen := len(cn.String())
+	// 8 bytes for the identity fingerprint.
+	for _, r := range fingerprint {
+		cnLen := len(cn.String())
+		runeLen := len(string(r))
+		if (cnLen + runeLen) > (prefixLen + 8) {
+			break
+		}
+		cn.WriteRune(r)
+	}
```

Comment:
Because we control the fingerprint format and it is base32, we do not need this loop here and can just take a slice.

Reply:
True. I simplified the implementation to assume that. Thank you. Resolving.